File: learnLeetCode/Day10/Part1/mst.py
"""
Day 10 — Part 1: Basic Connectivity (Minimum Spanning Tree)

Given n nodes and edges with weights, find minimum total cost to connect all nodes.

Approach: Kruskal's algorithm with Union-Find (Disjoint Set Union)
- Sort edges by weight
- For each edge (u, v, w), if u and v not yet connected, union them and add w to cost
- Time: O(E log E), Space: O(V + E)
"""
from typing import List


class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False
        if self.rank[px] < self.rank[py]:
            px, py = py, px
        self.parent[py] = px
        if self.rank[px] == self.rank[py]:
            self.rank[px] += 1
        return True


def minimum_spanning_tree(n: int, edges: List[List[int]]) -> int:
    """Returns minimum cost to connect all n nodes."""
    edges_sorted = sorted(edges, key=lambda e: e[2])
    uf = UnionFind(n)
    total_cost = 0
    edges_used = 0
    for u, v, w in edges_sorted:
        if uf.union(u, v):
            total_cost += w
            edges_used += 1
            if edges_used == n - 1:
                break
    # If we didn't use n-1 edges, graph is disconnected
    return total_cost if edges_used == n - 1 else -1


if __name__ == "__main__":
    # Example from problem
    n = 4
    edges = [[0,1,3],[0,2,5],[1,2,1],[2,3,4]]
    result = minimum_spanning_tree(n, edges)
    assert result == 8, f"Expected 8, got {result}"

    # Additional test
    assert minimum_spanning_tree(3, [[0,1,1],[1,2,2],[0,2,3]]) == 3
    assert minimum_spanning_tree(2, [[0,1,5]]) == 5

    print("Day10 Part1 mst.py: all tests passed")

File: learnLeetCode/Day10/Part1/mst.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Union-Find (Disjoint Set Union)
class UnionFind {
    vector<int> parent, rank;
public:
    UnionFind(int n) : parent(n), rank(n, 0) {
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    bool unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;
        if (rank[px] < rank[py]) swap(px, py);
        parent[py] = px;
        if (rank[px] == rank[py]) rank[px]++;
        return true;
    }
};

// Minimum Spanning Tree using Kruskal
int minimumSpanningTree(int n, vector<vector<int>> edges) {
    // Sort edges by weight
    sort(edges.begin(), edges.end(), [](auto &a, auto &b) { return a[2] < b[2]; });
    UnionFind uf(n);
    int totalCost = 0, edgesUsed = 0;
    for (auto &e : edges) {
        int u = e[0], v = e[1], w = e[2];
        if (uf.unite(u, v)) {
            totalCost += w;
            edgesUsed++;
            if (edgesUsed == n - 1) break;
        }
    }
    return (edgesUsed == n - 1) ? totalCost : -1;
}

int main() {
    // Example from problem
    int res1 = minimumSpanningTree(4, {{0,1,3},{0,2,5},{1,2,1},{2,3,4}});
    if (res1 != 8) return 1;

    // Additional tests
    if (minimumSpanningTree(3, {{0,1,1},{1,2,2},{0,2,3}}) != 3) return 1;
    if (minimumSpanningTree(2, {{0,1,5}}) != 5) return 1;

    cout << "Day10 Part1 mst.cpp: all tests passed" << endl;
    return 0;
}
